### ISSUE TITLE
Improve autosave performance

### ISSUE THREAD

**Issue by grz0** [wesnoth/wesnoth#5702]
### Game and System Information
 - What version of the game are you running?
a2c795bfd7a22c6196f2b5bd1386edaf1523e541 on Linux

### Describe the desired feature
While playing Wesnoth I've noticed that after the end of a turn game display remains frozen for about 2-3 seconds.
I was suspecting that this lag is caused by the autosave feature, so I disabled autosaves and the lag at the end of a turn disappeared.
I measured how much time is spend in `autosave_savegame::autosave()` and in `write_save_index()` (gzip compression was enabled):

~~~
20210413 02:11:10 info engine: Milliseconds to write_save_index(): 625
20210413 02:11:10 info engine: Milliseconds to write_save_index(): 629
20210413 02:11:10 info engine: Milliseconds to save DiD-Endless Night-Auto-Save30.gz: 1369
20210413 02:11:10 info engine: Deleting savegame 'DiD-Endless Night-Auto-Save24.gz'
20210413 02:11:11 info engine: Milliseconds to write_save_index(): 634
20210413 02:11:11 info engine: Milliseconds to autosave: 2008
~~~

`sum of time in write_save_index() = 1888 ms (avg 629 ms)`
`autosave time - write_save_index() = 120 ms`

If four `write_save_index()` calls could be replaced with only one at the end of `autosave_savegame::autosave()` then autosave time is reduced to:
`120 ms + 629 ms = 749 ms`
This is only 37% of the original autosave time.

In short, this feature request is about reducing the number of intermediate writes of the save index to the disk.  This would result in improved autosave performance. 

EDIT: please read comment 3.

**Comment by grz0** [comment]
I measured time also with disabled gzip compression:
~~~
20210413 00:34:57 info engine: Milliseconds to write_save_index(): 260
20210413 00:34:57 info engine: Milliseconds to write_save_index(): 273
20210413 00:34:57 info engine: Milliseconds to save DiD-Endless Night-Auto-Save27: 584
20210413 00:34:57 info engine: Deleting savegame '3p — Morituri-Auto-Save24'
20210413 00:34:57 info engine: Milliseconds to write_save_index(): 315
20210413 00:34:58 info engine: Milliseconds to write_save_index(): 339
20210413 00:34:58 info engine: Milliseconds to autosave: 1243
~~~


`sum of time in write_save_index() = 1187 ms (avg 297 ms)`
`autosave time - write_save_index() = 56 ms`
`297 ms + 56 ms = 353 ms`
Autosave time could be reduced to 30% of the original value.
